src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs

```diff
@@ -156,11 +184,14 @@ public void Dispose()
 
             DisposeOf(this.inputSubscriptions);
 
-            this.pipelineHead.Complete();
-            this.pipelineHead.Completion.Wait(TimeSpan.FromMilliseconds(this.pipelineConfiguration.PipelineCompletionTimeoutMsec));
+            TimeSpan pipelineDrainWaitTime = PollWaitForPipelineDrain();
+
+            pipelineHead.Complete();
+            TimeSpan completionWaitTime = TimeSpan.FromMilliseconds(Math.Max(100, this.pipelineConfiguration.PipelineCompletionTimeoutMsec - pipelineDrainWaitTime.TotalMilliseconds));
```

Comment:
I will add a comment. 

It is not just about draining the pipeline from events, but also about disposing of all internal resources that the pipeline blocks may use. Even if the pipeline is not fully drained (e.g. due to stuck filter/output), we still want to call Complete and give it some time.